### PR TITLE
Feature/fullscreen bar for X11

### PR DESCRIPTION
**Pull request by mmattes**
Here the PR for the floatbar for x11. This fixes Issue #4756 
What still needs to be verified is: 
- does it work in multimonitor mode?
- does it work with remote app, what means the floatbar should not be there at all?
- are the changes I made to the windows client correct so that the floatbar is deactivated if the new setting is used? 

Also let me know if any of my coding is bad as this is the first thing i really implemented using C

### CODE REVIEW

File: client/X11/resource/close.xbm [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#define close_width 24
+#define close_height 24
+static unsigned char close_bits[] = {
```
Comment (akallabeth) [inline client/X11/resource/close.xbm]: Formatting, run `./scripts/format_code.sh`

File: client/X11/resource/lock.xbm [hunk 2]
```diff
@@ -0,0 +1,9 @@
+#define lock_width 24
+#define lock_height 24
+static unsigned char lock_bits[] = {
```
Comment (akallabeth) [inline client/X11/resource/lock.xbm]: Formatting, run `./scripts/format_code.sh`

File: client/X11/resource/minimize.xbm [hunk 3]
```diff
@@ -0,0 +1,9 @@
+#define minimize_width 24
+#define minimize_height 24
+static unsigned char minimize_bits[] = {
+   0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+   0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+   0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x3f, 0x00, 0xfc,
```
Comment (akallabeth) [inline client/X11/resource/minimize.xbm]: Formatting, run `./scripts/format_code.sh`

File: client/X11/resource/restore.xbm [hunk 4]
```diff
@@ -0,0 +1,9 @@
+#define restore_width 24
+#define restore_height 24
+static unsigned char restore_bits[] = {
```
Comment (akallabeth) [inline client/X11/resource/restore.xbm]: Formatting, run `./scripts/format_code.sh`

File: client/X11/resource/unlock.xbm [hunk 5]
```diff
@@ -0,0 +1,9 @@
+#define unlock_width 24
+#define unlock_height 24
+static unsigned char unlock_bits[] = {
```
Comment (akallabeth) [inline client/X11/resource/unlock.xbm]: Formatting, run `./scripts/format_code.sh`

File: client/X11/xf_event.c [hunk 6]
```diff
@@ -1014,6 +1014,14 @@ BOOL xf_event_process(freerdp* instance, XEvent* event)
 		}
 	}
 
+	/* Forward event to the floatbar event handler, if the window
```
Comment (akallabeth) [inline client/X11/xf_event.c]: You should check the floatbar is actually used here before dispatching events

File: include/freerdp/settings.h [hunk 7]
```diff
@@ -1127,6 +1127,7 @@ struct rdp_settings
 	ALIGN64 BOOL   AsyncChannels;           /* 1546 */
 	ALIGN64 BOOL   AsyncTransport;          /* 1547 */
 	ALIGN64 BOOL   ToggleFullscreen;        /* 1548 */
+	ALIGN64 BOOL   Floatbar;
```
Comment (akallabeth) [inline include/freerdp/settings.h]: The numbers in the comments are there for a reason, to maintain ABI compatibility replace one of the reserved spots, do not just insert new variables

File: client/X11/xf_floatbar.c [hunk 8]
```diff
@@ -0,0 +1,609 @@
...
+}
+
+void xf_floatbar_hide_and_show(xfContext* xfc)
+{
+	xfFloatbar* floatbar;
+	floatbar = xfc->window->floatbar;
+
+	if (!floatbar->locked)
+	{
+		if (floatbar->mode == 0 && floatbar->last_motion_y_root > 10 && floatbar->y > (FLOATBAR_HEIGHT * -1))
```
Comment (akallabeth) [inline client/X11/xf_floatbar.c]: Might be a good idea to throw in some braces

**Comment by akallabeth** [comment]
Did some refactoring in https://github.com/akallabeth/FreeRDP/tree/pr/4805, you can cherry pick the commits from that branch as a start.

**Comment by mmattes** [comment]
@akallabeth thank you for your help and insights. Can you explain me why you moved the xfFloatbar struct from the header to the code file? Currently the hide doesnt work anymore and i need  to get your input on why you did that to fix it. 

Also please note that the event handling for the floatbar needs to happen before the main window handles its events and that i on purpose return the function with true afterwards. If i dont return the function then the events for the floatbar and its buttons are handled also by the main window, this has side effects. For example dragging the floatbar around also will cause a drag and drop action in the main window aswell.

**Comment by akallabeth** [comment]
Did move the stuff from the header as it is not used elsewhere (and shouldn't) so basically containment. As for event handling, ok, did not take that to account.

**Comment by mmattes** [comment]
@akallabeth are you happy the way it is now or is there still something i need to have a look at?

File: client/X11/xf_event.c [hunk 9]
```diff
@@ -366,6 +366,9 @@ static BOOL xf_event_MotionNotify(xfContext* xfc, XEvent* event, BOOL app)
 	if (xfc->use_xinput)
 		return TRUE;
 
+	if(xfc->floatbar && !(app))
```
Comment (akallabeth) [inline client/X11/xf_event.c]: This stuff is part of XEvent handling, couldn't that be done in that function?

File: client/X11/xf_window.c [hunk 10]
```diff
@@ -577,7 +577,7 @@ xfWindow* xf_CreateDesktopWindow(xfContext* xfc, char* name, int width,
 		            settings->DesktopPosY);
 	}
 
-	window->floatbar = xf_floatbar_new(xfc, window->handle, window->width);
+	window->floatbar = xf_floatbar_new(xfc, window->handle, xfc->workArea.width);
```
Comment (akallabeth) [inline client/X11/xf_window.c]: This must be the center of the screen that the window is actually on (which might change on multimonitor systems)

**Comment by akallabeth** [comment]
@mmattes Tested your pull, looks good except the wrong positioning.
Would suggest you clean that up and stash all changes in a single commit.

File: client/X11/xf_event.c (hunk 9 above)
Comment (mmattes) [reply client/X11/xf_event.c]: I need to know in the floatbar where the mouse pointer is within the root window to decide wether or not the floatbar needs to be shown or hidden. That value is checked outside of any Event handling as the main loop calls the xf_floatbar_hide_and_show function regulary. 
Alternativeley i could store the last x, y postion of the pointer outside of my xfFloatbar struct, but i did not want to mess around to much with the existing structures.

Comment (akallabeth) [reply client/X11/xf_event.c]: @mmattes Ah, ok, then that is a valid case.

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by akallabeth** [comment]
@mmattes sorry, did take some time to check this.

**Comment by mmattes** [comment]
@akallabeth cool, i will then create a new branch and PR as this branch is somehow messed up when i tried to squash the changes into one commit.

**Comment by akallabeth** [comment]
@mmattes Looks like you have some clash with redefinitions in the windows client.
Might also be a good idea to refactor the commit history and clean that up before we merge.

**Comment by akallabeth** [comment]
@mmattes Ok, forget the last comment, yours just showed up above mine (damn auto refresh, or lack thereof sometimes)

**Comment by mmattes** [comment]
@akallabeth can you point out to me what the problem is in regards to the windows client? I just compiled it on a windows machine and it looks like everything works?

**Comment by akallabeth** [comment]
Struct member floatbar redefinition see https://ci.freerdp.com/job/windows_msbuild_vs2010/1220/arch=win64,label=vs2010/console for the build log

**Comment by mmattes** [comment]
thank you will have a look at it, couldnt find the log in jenkins as i am not used to it.

**Comment by mmattes** [comment]
im closing this see #4865

**Pull request closed without merging**